app/src/main/java/org/jellyfin/mobile/webapp/ConnectionHelper.kt

```diff
@@ -0,0 +1,207 @@
+    }
+
+    fun onSelectServer() {
+        cachedInstanceUrl = null
+        loadOrShowSetup()
+    }
+
+    fun onErrorReceived() {
+        connected = false
+        showConnectionError()
+        onSelectServer()
+    }
+
+    fun onBackPressed(): Boolean {
+        if (serverSetupLayout.isAttachedToWindow) {
+            rootView.removeView(serverSetupLayout)
+            cachedInstanceUrl = appPreferences.instanceUrl?.toHttpUrlOrNull()
+            webView.isVisible = true
+            return true
+        }
+        return false
+    }
+
+    private fun loadOrShowSetup() {
+        cachedInstanceUrl.let { url ->
+            if (url != null) {
+                webView.isVisible = true
+                webView.loadUrl(url.resolve(Constants.INDEX_PATH).toString())
+            } else {
+                webView.isVisible = false
+                showServerSetup()
+            }
+        }
+    }
+
+    private fun showServerSetup() {
+        rootView.addView(serverSetupLayout)
+        hostInput.setText(appPreferences.instanceUrl)
+        hostInput.setSelection(hostInput.length())
+        hostInput.setOnEditorActionListener { _, action, event ->
+            when {
+                action == EditorInfo.IME_ACTION_DONE || event.keyCode == KeyEvent.KEYCODE_ENTER -> {
+                    connect()
+                    true
+                }
+                else -> false
+            }
+        }
+        connectButton.setOnClickListener {
+            connect()
+        }
+
+        // Show keyboard
+        serverSetupLayout.doOnNextLayout {
+            hostInput.postDelayed(25) {
+                hostInput.requestFocus()
+                activity.getSystemService<InputMethodManager>()?.showSoftInput(hostInput, InputMethodManager.SHOW_IMPLICIT)
+            }
+        }
+    }
+
+    private fun connect() {
+        hostInput.isEnabled = false
+        connectButton.isEnabled = false
+        clearConnectionError()
+        activity.lifecycleScope.launch {
+            val httpUrl = checkServerUrlAndConnection(hostInput.text.toString())
+            if (httpUrl != null) {
+                appPreferences.instanceUrl = httpUrl.toString()
+                cachedInstanceUrl = httpUrl
+                rootView.removeView(serverSetupLayout)
+                loadOrShowSetup()
+            }
+            hostInput.isEnabled = true
+            connectButton.isEnabled = true
+        }
+    }
+
+    private fun showConnectionError(@StringRes errorString: Int = R.string.connection_error_cannot_connect) {
+        connectionErrorText.setText(errorString)
+        connectionErrorText.isVisible = true
+    }
+
+    private fun clearConnectionError() {
+        connectionErrorText.isVisible = false
+    }
+
+    private suspend fun checkServerUrlAndConnection(enteredUrl: String): HttpUrl? {
+        val normalizedUrl = enteredUrl.run {
+            if (lastOrNull() == '/') this
+            else "$this/"
+        }
+
+        val urls = when {
```

Comment:
We should change this part to use the helper method from the apiclient in the future.

Reply:
Definitely, yes.